File: two_pointer/dutch-national-flag.js
const dutch_flag_sort = (arr) => {
    let l = 0;
    let r = arr.length - 1;

    let i = 0;
    while (i <= r) {
        if (arr[i] === 0) {
            [arr[i], arr[l]] = [arr[l], arr[i]];
            l++;
            i++;
        } else if (arr[i] === 1) {
            i++;
        } else {
            [arr[i], arr[r]] = [arr[r], arr[i]];
            r--;
        }
    }

    return arr;
}

console.log(dutch_flag_sort([1, 0, 2, 1, 0])) // [0 0 1 1 2]
console.log(dutch_flag_sort([2, 2, 0, 1, 2, 0])) // [0 0 1 2 2 2]

File: two_pointer/python/dutch-national-flag.py
def dutch_flag_sort(arr):
    lz, ft = -1, len(arr)
    i = 0
    while i < ft:
        if arr[i] == 0 and i - lz > 1:
            lz += 1
            arr[i], arr[lz] = arr[lz], arr[i]
        elif arr[i] == 2 and ft - i > 1:
            ft -= 1
            arr[i], arr[ft] = arr[ft], arr[i]
        else:
            i += 1

    return arr


if __name__ == '__main__':
    print(dutch_flag_sort([1, 0, 2, 1, 0]))
    print(dutch_flag_sort([2, 2, 0, 1, 2, 0]))
